README.md
# Statistical Modelling of AP-MS Data (SMAD)

This R package implements statistical modelling of affinity purification–mass spectrometry (AP-MS) data to compute confidence scores to identify *bona fide* protein-protein interactions (PPI).

## Installation

The development version can be installed through github:
```{r}
 devtools::install_github(repo="zqzneptune/SMAD")
 library(SMAD)
```
## Input Data
A demo data.frame was provided as a hint how the input data should strcutured in order to run the scoring functions:

```{r}
data(TestDatInput)
colnames(TestDataInput)

[1] "idRun" "idBait" "idPrey" "countPrey" "lenPrey" 
```

|idRun|idBait|idPrey|countPrey|lenPrey|
|-----|:----:|:----:|:-------:|-------|
|Unique ID of one AP-MS run|Bait ID|Prey ID|Prey peptide count|Protein sequence length of the prey|

In case of duplcates, a suffix or prefix of e.g. "A", "B" could be added to **idRun** in order to make **"idRun-idBait"** combination unique to each relicate.

## Run scoring

### 1. CompPASS

Comparative Proteomic Analysis Software Suite (CompPASS) is based on spoke model. This algorithm was developed by Dr. Mathew Sowa for defining the human deubiquitinating enzyme interaction landscape [(Sowa, Mathew E., et al., 2009)][1]. The implementation of this algorithm was inspired by Dr. Sowa's [online tutorial][2]. The output includes Z-score, S-score, D-score and WD-score. In its implementation in BioPlex 1.0 [(Huttlin, Edward L., et al., 2015)][3] and 
BioPlex 2.0 [(Huttlin, Edward L., et al., 2017)][4], a naive 
Bayes classifier that learns to distinguish true interacting proteins from 
non-specific background and false positive identifications was included in the 
compPASS pipline. This function was optimized from the [source code][5].

The input data.frame, *datInput*, should include:**idRun**, **idBait**, **idPrey** and **countPrey**.

```{r}
datScore <- CompPASS(datInput)
```

### 2. DICE

The Dice coefficient is used to score the interaction scores across prey pair-wise combinations, which was proposed by [(Bing Zhang et al., 2008)][9]

The input data.frame, *datInput*, should include:**idRun** and **idPrey**.

```{r}
datScore <- DICE(datInput)
```

### 3. Hart

Hart scoring algorithm is based on a hypergeometric distribution error model [(Hart et al., 2007)][6].

The input data.frame, *datInput*, should include:**idRun** and **idPrey**.

```{r}
datScore <- Hart(datInput)
```


### 4. HGScore

HGScore algorithm is based on a hypergeometric distribution error model [(Hart et al., 2007)][6] with incorporation of NSAF [(Zybailov, Boris, et al., 2006)][7]. This algorithm was first introduced to predict the protein complex network of Drosophila melanogaster [(Guruharsha, K. G., et al., 2011)][8]. This scoring algorithm was based on matrix model. 

The input data.frame, *datInput*, should include:**idRun**, **idPrey**, **countPrey** and **lenPrey**.

```{r}
datScore <- HG(datInput)
```

### 5. PE
PE incorporated both spoke and matrix model as repored in [(Sean R. Collins, et al., 2007)][10].

The input data.frame, *datInput*, should include:**idRun**, **idBait** and **idPrey**.

```{r}
datScore <- PE(datInput)
```



## License

MIT @ Qingzhou Zhang

[1]: https://doi.org/10.1016/j.cell.2009.04.042
[2]: http://besra.hms.harvard.edu/ipmsmsdbs/cgi-bin/tutorial.cgi
[3]: https://doi.org/10.1016/j.cell.2015.06.043
[4]: https://www.nature.com/articles/nature22366
[5]: https://github.com/dnusinow/cRomppass
[6]: https://doi.org/10.1186/1471-2105-8-236
[7]: https://doi.org/10.1021/pr060161n
[8]: https://doi.org/10.1016/j.cell.2011.08.047
[9]: https://doi.org/10.1093/bioinformatics/btn036
[10]: https://doi.org/10.1074/mcp.M600381-MCP200

NEWS
# SMAD 1.3.1

* Add PE() function combing Spoke and Matrix model
# SMAD 0.1.0

## New functions

* Spoke model `CompPASS()` function.

* Matrix model `HG()` function.

## Improved speed by Rcpp computing


R/HG.R
#' HGScore
#' Scoring algorithm based on a hypergeometric distribution error model 
#' (Hart et al.,2007) with incorporation of NSAF (Zybailov, Boris, et al., 2006)
#' . This algorithm was first introduced to predict the protein complex network 
#' of Drosophila melanogaster (Guruharsha, K. G., et al., 2011). This scoring 
#' algorithm was based on matrix model.
#'
#' @title HGScore
#' @param datInput A dataframe with column names: idRun, idPrey, countPrey, 
#' lenPrey. Each row represent one unique protein captured in one pull-down 
#' experiment.
#'
#' @return A dataframe consists of pairwise combindation of preys identified in 
#' the input with HG scores indicating interacting probabilities computed from 
#' negative log transformed Hypergeometric test P-values.
#'
#' @author Qingzhou Zhang, \email{zqzneptune@hotmail.com}
#' @references Guruharsha, K. G., et al. 'A protein complex network of 
#' Drosophila melanogaster.' Cell 147.3 (2011): 690-703.
#' \url{https://doi.org/10.1016/j.cell.2011.08.047}
#' @references Hart, G. Traver, Insuk Lee, and Edward M. Marcotte. 
#' 'A high-accuracy consensus map of yeast protein complexes reveals modular 
#' nature of gene essentiality.' BMC bioinformatics 8.1 (2007): 236.
#' \url{https://doi.org/10.1186/1471-2105-8-236}
#' @references Zybailov, Boris, et al. 'Statistical analysis of membrane 
#' proteome expression changes in Saccharomyces c erevisiae.' Journal of 
#' proteome research 5.9 (2006): 2339-2347.
#' \url{https://doi.org/10.1021/pr060161n}

#' @importFrom dplyr group_by
#' @importFrom dplyr summarise
#' @importFrom dplyr mutate
#' @importFrom dplyr left_join
#' @importFrom dplyr filter
#' @importFrom dplyr n
#' @importFrom tidyr spread
#' @importFrom magrittr %>%
#' @importFrom utils combn
#' @importFrom dplyr bind_rows
#' @importFrom stats setNames
#' @importFrom stats phyper
#' @import Rcpp
#' @importFrom Rcpp evalCpp
#' @useDynLib SMAD
#' @exportPattern '^[[:alpha:]]+'
#' @export
#' @examples
#' data(TestDatInput)
#' datScore <- HG(TestDatInput)
#' head(datScore)


HG <- function(datInput) {
    colInput <-
        c("idRun", "idPrey", "countPrey", "lenPrey")
    
    if(!is.data.frame(datInput)){
        stop("Input data should be data.frame")
    }
    
    if(!all(colInput %in% colnames(datInput))){
        missingCol <-
            setdiff(colInput, 
                    colnames(datInput)[match(colInput, colnames(datInput))])
        stop("Input data missing: ", paste(missingCol, collapse = ", "))
    }
    
    
    . <- NULL
    idRun <- NULL
    countPrey <- NULL
    lenPrey <- NULL
    NormalSpec <- NULL
    SumNS <- NULL
    NSAF <- NULL
    NormalNSAF <- NULL
    Tn <- NULL
    UniprotID <- NULL
    tnA <- NULL
    tnB <- NULL
    NMinTn <- NULL
    HG <- NULL
    ppiTN <- NULL
    datCnt <- 
        datInput %>% 
        mutate(`NormalSpec` = `countPrey`/`lenPrey`) %>% 
        group_by(`idRun`) %>% 
        mutate(`SumNS` = sum(`NormalSpec`)) %>% 
        mutate(`NSAF` = `NormalSpec`/`SumNS`) %>% 
        group_by(`idRun`) %>% 
        mutate(`NormalNSAF` = `NSAF`/min(`NSAF`)) %>% 
        mutate(`Tn` = as.integer(sqrt(`NormalNSAF`)))
    d <- spread(datCnt[, c("idRun", "idPrey", "Tn")], 
                `idRun`, `Tn`)
    g <- as.matrix(d[, -1])
    g[is.na(g)] <- 0
    rownames(g) <- d$idPrey
    pps <- combn(d$idPrey, 2)
    PPN <- .GetPPN(t(g))
    CppPPN <- PPN[lower.tri(PPN, diag = FALSE)]
    datPPI <- data.frame(cbind(t(pps[, CppPPN != 0]), 
                                CppPPN[CppPPN != 0]), stringsAsFactors = FALSE)
    colnames(datPPI) <- 
        c("InteractorA", "InteractorB", "ppiTN")
    datPPI$ppiTN <- 
        as.numeric(datPPI$ppiTN)
    tnInteractorA <- 
        datPPI[, c("InteractorA", "ppiTN")]
    colnames(tnInteractorA) <- 
        c("UniprotID", "ppiTN")
    tnInteractorB <- 
        datPPI[, c("InteractorB", "ppiTN")]
    colnames(tnInteractorB) <- 
        c("UniprotID", "ppiTN")
    tnProtein <- 
        bind_rows(tnInteractorA, tnInteractorB) %>% 
        group_by(`UniprotID`) %>% 
        summarise(minTn = sum(`ppiTN`))
    sumMinTnInteractorA <- 
        tnProtein
    colnames(sumMinTnInteractorA) <- 
        c("InteractorA", "tnA")
    sumMinTnInteractorB <- 
        tnProtein
    colnames(sumMinTnInteractorB) <- 
        c("InteractorB", "tnB")
    scorePPI <- 
        datPPI %>% 
        left_join(., sumMinTnInteractorA, by = "InteractorA") %>% 
        left_join(., sumMinTnInteractorB, by = "InteractorB") %>% 
        mutate(`NMinTn` = sum(tnProtein$minTn)/2) %>% 
        mutate(`HG` = -phyper(`ppiTN`, `tnA`, `NMinTn` - `tnB`, 
                                `tnB`, lower.tail = FALSE, log.p = TRUE))
    return(scorePPI)
}
